systems-architecture.md

```diff
@@ -32,23 +32,42 @@ We wish to ensure the following in our implementation:
 
-#### Registration session
+### Opening and using a registration session
 
-1. The client authenticates the key server and opens a channel with the key server as specified [above](#underlying_transport_layer).
-1. The client and key server run the OPAQUE registration stage; see [this section](current-development-phase.md#cryptographic-protocol-and-implementation-dependencies) for protocol version and dependency information.
+1. The client authenticates the key server and opens a channel with the key server as specified in the [underlying transport section](#systems-architecture.md/#underlying_transport_layer).
+1. The client and key server run the OPAQUE registration stage; see [this section](current-development-phase.md#cryptographic-protocol-and-implementation-dependencies) for protocol version and dependency information. The implementation must satisfy the following:
+    1. The key server must check that the given user identifier `user_id` is _fresh_ (i.e., no user has previously registered with the given identifier) and of the expected format and length. If this check fails, the key server MUST fail the registration request.
     1. The client receives as output a value `export_key`, which is a pseudorandom value, independent of all other OPAQUE protocol values, that is known only to the client.
-    1. The key server receives as output a record that corresponds to the client's registration.
+    1. The key server receives as output a record that corresponds to the client's registration, which includes an identifier `user_id` that matches that of the client.
+1. The client and key server mutually authenticate via the OPAQUE protocol's authenticated key exchange (AKE); see [this section](current-development-phase.md#cryptographic-protocol-and-implementation-dependencies) for protocol version and dependency information. The implementation must satisfy the following:
+    1. The client receives as output two values, an `export_key` (matching that from the registration phase) and a `session_key` (which is the output of an authenticated key exchange).
+    1. The server receives as output a `session_key` matching that of the client.
+    1. Both client and server MUST receive confirmation of the success of the AKE.
+1. The client and server open an encrypted channel secured under a key derived from `session_key`. [TODO](https://github.com/boltlabs-inc/key-mgmt-spec/issues/29): Include additional details here.
+1. At this point, the registration session is considered _open_. 
+1. We have the following requirements for using an open registration session:
+    1. All messages sent between the client and server MUST be over this encrypted channel, i.e., all messages should be encrypted under the given shared key. 
+    1. The key server MUST additionally reject all messages sent over this channel that fail the AEAD verification checks.
+    1. The only valid request for a registration is a request to [complete registration](TODO: fill me in).
```

Comment:
What does this mean? I thought at this point we could go on to either terminate the session or make a request (store, retrieve, audit, import, export) as below. I thought registration was already complete at this stage.

Reply:
So to complete registration we have to generated the storage_key and store it at the server. I may have been inconsistent in my draft on how to write about this.

Reply:
L18 in register.md is talking about what "complete registration" requests are.

Reply:
Oh, I see. Fixed the link.